mvrx/src/main/kotlin/com/airbnb/mvrx/CoroutinesStateStore.kt

```diff
@@ -0,0 +1,85 @@
+package com.airbnb.mvrx
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ConflatedBroadcastChannel
+import kotlinx.coroutines.channels.consumeEach
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.AtomicInteger
+
+@Suppress("EXPERIMENTAL_API_USAGE")
+class CoroutinesStateStore<S : MvRxState>(
+    initialState: S,
+    private val scope: CoroutineScope = createCoroutineScope()
+) : MvRxStateStore<S> {
+
+    private val setStateChannel = Channel<S.() -> S>(capacity = Channel.UNLIMITED)
+    private val withStateChannel = Channel<(S) -> Unit>(capacity = Channel.UNLIMITED)
+    private val flushQueuesChannel = Channel<Unit>(capacity = Channel.CONFLATED)
+
+    private val stateChannel = ConflatedBroadcastChannel<S>(initialState)
+    override val state: S get() = stateChannel.valueOrNull ?: error("There is no state yet! This should never happen.")
+
+    init {
+        scope.launch {
+            flushQueuesChannel.consumeEach {
```

Comment:
I was thinking about how we could simplify this. I think we could do something like this
```
sealed class StateAction
class SetState(reducer) : StateAction()
class GetState(block) : StateAction()

private val stateActionChannel = PriorityChannel<StateAction>(comparator = { // state actions get priority }
```

A bit of psuedo code, but any way, leverage sealed classes to unify into a single channel and then use an idea like PriorityQueue to poll and make sure we get all SetState actions first.

Main problem is coroutines don't seem to provide something out of the box like "PriorityChannel" although this thread gave me some ideas https://github.com/Kotlin/kotlinx.coroutines/issues/238 and this library has an implementation 
https://github.com/kerubistan/kroki/blob/master/kroki-coroutines/src/main/kotlin/io/github/kerubistan/kroki/coroutines/Channels.kt

Since we probably don't want to implement our own, I think going with the `select` approach here would be nice https://github.com/Kotlin/kotlinx.coroutines/issues/238#issuecomment-368080421

Then we still need a separate channel for get/set, but we can get rid of the flush channel and just select between get/set

Reply:
@elihart That seems slightly less optimized because now we'd have to create a container object for each action and the queue would have to be sorted. I'm also not sure it would be easier to read. What do you think?